### PR TITLE
feat: add orphan resource cleanup during reconcile

### PR DESCRIPTION
**Pull request by yizhaodev**
## Description

Previously, when a CR spec change removed child resources (e.g. disabling grafana or monitoring), the stale resources remained in the cluster. The controller now diffs rendered objects against existing owned resources after each reconcile and deletes orphans. Missing CRDs are skipped gracefully.

## How Has This Been Tested?

`make test` — two new envtest tests:

- Disabling grafana on a CR deletes the dashboard ConfigMap, other resources unaffected
- Disabling grafana on one CR does not affect another CR's resources

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

## Release Notes

* **New Features**
  * Added LLMBatchGateway specification validation before rendering with clear error reporting for configuration issues.
  * Implemented automatic cleanup of orphaned managed resources to prevent resource leaks.
  * Added support for digest-based container image references in addition to tag-based references.

* **Breaking Changes**
  * Removed namespace field from secret references; secrets must now be located in the operator namespace.

* **Documentation**
  * Added Design Document link to README.

### CODE REVIEW

File: internal/controller/llmbatchgateway_controller.go [hunk 1]
```diff
@@ -87,13 +108,80 @@ func (r *LLMBatchGatewayReconciler) Reconcile(ctx context.Context, req ctrl.Requ
...
+			Version: gvk.Version,
+			Kind:    gvk.Kind + "List",
+		})
+
+		if err := r.List(ctx, &list, client.InNamespace(gw.Namespace)); err != nil {
+			if apierrors.IsNotFound(err) || meta.IsNoMatchError(err) {
+				logger.V(1).Info("skipping orphan check (CRD not installed)", "kind", gvk.Kind)
+				continue
+			}
+			errs = append(errs, fmt.Errorf("listing %s: %w", gvk.Kind, err))
```
[the remark this answers is not in the document]
Comment (yizhaodev) [reply internal/controller/llmbatchgateway_controller.go:172]: Good point about server-side filtering. However, the Helm chart's labels are not consistent across all resource types — PrometheusRule and Grafana dashboard ConfigMap lack `app.kubernetes.io/instance`. Fixing this requires upstream template changes. At the current scale (~10 resources per CR), client-side filtering has no measurable impact. Will revisit if the resource count grows.

File: internal/controller/llmbatchgateway_controller.go [hunk 2]
```diff
@@ -87,13 +108,80 @@ func (r *LLMBatchGatewayReconciler) Reconcile(ctx context.Context, req ctrl.Requ
...
+			continue
+		}
+
+		for i := range list.Items {
+			item := &list.Items[i]
+			if !isOwnedBy(item, gw) {
+				continue
```
[the remark this answers is not in the document]
Comment (yizhaodev) [reply internal/controller/llmbatchgateway_controller.go:179]: Agreed. Added `isControllerOwnedBy` that checks UID + Kind + APIVersion + Controller=true, used only in the cleanup path. The looser `isOwnedBy` is kept for `updateStatus` where it's sufficient.

File: internal/controller/llmbatchgateway_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yizhaodev) [reply internal/controller/llmbatchgateway_controller.go:172]: Good idea on the partial optimization. Will consider that when we revisit.

File: test/e2e/e2e_test.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	for _, c := range conditions {
+		if typ, ok := c["type"].(string); ok {
+			found[typ] = true
+		}
+	}
+	for _, e := range expected {
+		if !found[e] {
+			t.Errorf("missing condition %q", e)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (yizhaodev) [reply test/e2e/e2e_test.go:44]: fixed, thanks

File: test/e2e/helpers_test.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+func kubectlGetExists(t *testing.T, resource, name, namespace string) bool {
+	t.Helper()
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	_, err := kubectl(ctx, "get", resource, name, "-n", namespace, "--no-headers")
+	return err == nil
+}
```
[the remark this answers is not in the document]
Comment (yizhaodev) [reply test/e2e/helpers_test.go:53]: fixed, thanks

**Comment by lioraron** [comment]
Good work overall — orphan cleanup is correct, well-tested, and handles edge cases properly. Three comments:

1. **`test/e2e/helpers_test.go` — `waitForCondition` is unused dead code.**
   It's defined but never called. `testStatusConditions` implements its own polling loop with different semantics (checks all three conditions are present vs. `waitForCondition` which checks one condition's status value). Either use it or remove it.

2. **`internal/controller/helm.go:323` — `splitImage` digest handling relies on chart template format.**
   Splitting `repo@sha256:abc` into `("repo@sha256", "abc")` depends on the chart template doing `printf "%s:%s"` to reconstruct the original string. If the chart template ever changes that format string, this breaks silently. The tests document the contract which mitigates the risk, but worth a brief comment noting the chart-template dependency.

3. **`internal/controller/llmbatchgateway_controller.go:33` — `managedGVKs` must stay in sync with the chart.**
   This list must be a superset of all GVK types the chart can produce — if a new type is added to the chart (e.g. NetworkPolicy, HPA) but not here, orphans of that type won't be cleaned up. Worth a short comment noting this invariant so future chart changes remember to update it.

_(Posting as a top-level comment due to GitHub incident affecting line comments — will convert to inline comments when resolved.)_

**Comment by yizhaodev** [comment]
> Good work overall — orphan cleanup is correct, well-tested, and handles edge cases properly. Three comments:
> 
> 1. **`test/e2e/helpers_test.go` — `waitForCondition` is unused dead code.**
>    It's defined but never called. `testStatusConditions` implements its own polling loop with different semantics (checks all three conditions are present vs. `waitForCondition` which checks one condition's status value). Either use it or remove it.
> 2. **`internal/controller/helm.go:323` — `splitImage` digest handling relies on chart template format.**
>    Splitting `repo@sha256:abc` into `("repo@sha256", "abc")` depends on the chart template doing `printf "%s:%s"` to reconstruct the original string. If the chart template ever changes that format string, this breaks silently. The tests document the contract which mitigates the risk, but worth a brief comment noting the chart-template dependency.
> 3. **`internal/controller/llmbatchgateway_controller.go:33` — `managedGVKs` must stay in sync with the chart.**
>    This list must be a superset of all GVK types the chart can produce — if a new type is added to the chart (e.g. NetworkPolicy, HPA) but not here, orphans of that type won't be cleaned up. Worth a short comment noting this invariant so future chart changes remember to update it.
> 
> _(Posting as a top-level comment due to GitHub incident affecting line comments — will convert to inline comments when resolved.)_

Thanks for the review!
1. `waitForCondition` unused: will remove it.
2. `splitImage` chart-template dependency: have to fix helm template from upstream first to support digest
3. `managedGVKs` sync invariant: will add a comment noting it must be a superset of chart-produced GVKs.

File: internal/controller/llmbatchgateway_controller.go [hunk 5]
```diff
@@ -55,6 +78,22 @@ func (r *LLMBatchGatewayReconciler) Reconcile(ctx context.Context, req ctrl.Requ
...
+				Reason:             "ValidationFailed",
+				Message:            err.Error(),
+				ObservedGeneration: gw.Generation,
+			})
+		}
+		if statusErr := r.Status().Update(ctx, &gw); statusErr != nil {
+			logger.Error(statusErr, "failed to update status after validation failure")
+		}
+		return ctrl.Result{}, nil
+	}
```
Comment (lioraron) [inline internal/controller/llmbatchgateway_controller.go:95]: **Bug:** If `r.Status().Update()` fails here, the error is logged but `nil` is returned, so the controller-runtime won't requeue. The user's CR will have stale/absent conditions with no signal that validation failed, and the controller won't retry the status update until the user makes another spec change.

The existing `RenderChart` failure path (line ~101) has the same structure but returns the render error (`fmt.Errorf("rendering chart: %w", err)`), which triggers a requeue. Here, the validation error is permanent and shouldn't be returned (spec won't change without user action), but the *status update* error is transient and should trigger a requeue:

```go
if statusErr := r.Status().Update(ctx, &gw); statusErr != nil {
    return ctrl.Result{}, fmt.Errorf("updating status after validation failure: %w", statusErr)
}
return ctrl.Result{}, nil
```

Comment (yizhaodev) [reply internal/controller/llmbatchgateway_controller.go:95]: thanks, this is fiex.

**Comment by yizhaodev** [comment]
@lioraron all comments have been addressed, please review again

**Review by lioraron — APPROVED** [review]
